src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfTPrimitive.cs

```diff
@@ -13,18 +13,17 @@ namespace System.Text.Json.Nodes
     internal sealed class JsonValuePrimitive<TValue> : JsonValue<TValue>
     {
         private readonly JsonConverter<TValue> _converter;
-        private readonly JsonValueKind _valueKind;
+        private JsonValueKind? _valueKind;
 
         public JsonValuePrimitive(TValue value, JsonConverter<TValue> converter, JsonNodeOptions? options) : base(value, options)
         {
             Debug.Assert(TypeIsSupportedPrimitive, $"The type {typeof(TValue)} is not a supported primitive.");
             Debug.Assert(converter is { IsInternalConverter: true, ConverterStrategy: ConverterStrategy.Value });
 
             _converter = converter;
-            _valueKind = DetermineValueKind(value);
         }
 
-        private protected override JsonValueKind GetValueKindCore() => _valueKind;
+        private protected override JsonValueKind GetValueKindCore() => _valueKind ??= DetermineValueKind(Value);
```

Comment:
Nullable<T> assignment is not atomic and thread safe. Is it guaranteed that there are no multiple threads running through this in parallel?

Reply:
Good catch. The type is mutable and not threadsafe but we should guarantee that concurrent reads are safe.